File: app_mobile/kmp/androidApp/src/main/java/com/ansmi/gestsquadre/kmp/push/FcmSessionRegistry.kt
package com.ansmi.gestsquadre.kmp.push

import com.ansmi.gestsquadre.shared.model.SquadSession
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

/** Collega onNewToken (service) alla sessione attiva dopo login. */
object FcmSessionRegistry {
    @Volatile
    var session: SquadSession? = null

    @Volatile
    var scope: CoroutineScope? = null

    @Volatile
    var onToken: (suspend (SquadSession, String) -> Unit)? = null

    fun bind(
        activeSession: SquadSession,
        coroutineScope: CoroutineScope,
        register: suspend (SquadSession, String) -> Unit,
    ) {
        session = activeSession
        scope = coroutineScope
        onToken = register
    }

    fun clear() {
        session = null
        scope = null
        onToken = null
    }

    fun deliverNewToken(token: String) {
        val active = session ?: return
        val register = onToken ?: return
        val coroutineScope = scope ?: return
        coroutineScope.launch {
            runCatching { register(active, token) }
        }
    }
}

File: app_mobile/kmp/iosApp/iosApp/FcmSessionRegistry.swift
import Foundation
import shared

/// Collega onNewToken (Firebase) alla sessione attiva dopo login.
final class FcmSessionRegistry {
    static let shared = FcmSessionRegistry()

    private var session: SquadSession?
    private var onToken: ((SquadSession, String) -> Void)?

    private init() {}

    func bind(session: SquadSession, onToken: @escaping (SquadSession, String) -> Void) {
        self.session = session
        self.onToken = onToken
    }

    func clear() {
        session = nil
        onToken = nil
    }

    func deliverNewToken(_ token: String) {
        guard let session, let onToken else { return }
        onToken(session, token)
    }
}
